File: client/dexterity/dummy_oracle/instructions/update_clock.py
from solana.publickey import PublicKey
from solana.transaction import AccountMeta, TransactionInstruction
from podite import I64, U64, pod

from .common import InstructionCode
from dexterity.program_ids import ORACLE_PROGRAM_ID


@pod
class Params:
    instr: InstructionCode
    slot: U64
    epoch_start_timestamp: I64
    epoch: U64
    leader_schedule_epoch: U64
    unix_timestamp: I64


def update_clock_ix(
    clock: PublicKey,
    slot: int,
    epoch_start_timestamp: int,
    epoch: int,
    leader_schedule_epoch: int,
    unix_timestamp: int,
):
    keys = [
        AccountMeta(pubkey=clock, is_signer=False, is_writable=True),
    ]
    params = Params(
        instr=InstructionCode.UPDATE_CLOCK,
        slot=slot,
        epoch_start_timestamp=epoch_start_timestamp,
        epoch=epoch,
        leader_schedule_epoch=leader_schedule_epoch,
        unix_timestamp=unix_timestamp,
    )
    return TransactionInstruction(
        keys=keys, program_id=ORACLE_PROGRAM_ID, data=params.to_bytes()
    )

File: dexteritysdk/src/oracle/update_clock.rs
use crate::{
    common::utils::{SDKError, SDKResult},
    sdk_client::{ClientSubset, SDKClient},
    state::clone_keypair,
    KeypairD, SDKContext,
};
use solana_program::{
    instruction::Instruction, message::Message, program_error::ProgramError,
    program_option::COption, program_pack::Pack, pubkey::Pubkey, sysvar::rent::Rent,
};
use solana_sdk::{
    account::Account,
    client::{Client, SyncClient},
    signature::{Keypair, Signer},
    system_instruction,
    transaction::Transaction,
    transport::TransportError,
};

pub async fn update_clock_account(
    client: &SDKClient,
    dummy_oracle_program_id: Pubkey,
    authority: &KeypairD,
    system_program_id: Pubkey,
    slot: u64,
    epoch_start_timestamp: i64,
    epoch: u64,
    leader_schedule_epoch: u64,
    unix_timestamp: i64,
) -> std::result::Result<Pubkey, SDKError> {
    let seeds: Vec<Vec<u8>> = vec![b"clock".to_vec()];
    let (clock_metadata_key, _) = Pubkey::find_program_address(
        seeds
            .iter()
            .map(|v| v.as_slice())
            .collect::<Vec<&[u8]>>()
            .as_ref(),
        &dummy_oracle_program_id,
    );
    let initialize_clock_ix = dummy_oracle::processor::update_clock_ix(
        dummy_oracle_program_id,
        clock_metadata_key,
        authority.pubkey(),
        system_program_id,
        dummy_oracle::processor::update_clock::Params {
            slot: slot,
            epoch_start_timestamp: epoch_start_timestamp,
            epoch: epoch,
            leader_schedule_epoch: leader_schedule_epoch,
            unix_timestamp: unix_timestamp,
        },
    );
    client
        .sign_send_instructions(vec![initialize_clock_ix], vec![authority])
        .await?;
    Ok(clock_metadata_key)
}
